src/Internal/ProductAttributesLookup/LookupDataStore.php

```diff
@@ -0,0 +1,301 @@
+			return $term_ids_by_slug_cache[ $taxonomy ][ $term_slug ];
+		} else {
+			return null;
+		}
+	}
+
+	/**
+	 * Get the variations of a given variable product.
+	 *
+	 * @param \WC_Product_Variable $product The product to get the variations for.
+	 * @return array An array of WC_Product_Variation objects.
+	 */
+	private function get_variations_of( \WC_Product_Variable $product ) {
+		$variation_ids = $product->get_children();
+		return array_map(
+			function( $id ) {
+				return WC()->call_function( 'wc_get_product', $id );
+			},
+			$variation_ids
+		);
+	}
+
+	/**
+	 * Check if a given product is a variable product.
+	 *
+	 * @param \WC_Product $product The product to check.
+	 * @return bool True if it's a variable product, false otherwise.
+	 */
+	private function is_variable_product( \WC_Product $product ) {
+		return is_a( $product, \WC_Product_Variable::class );
+	}
+
+	/**
+	 * Check if a given product is a variation.
+	 *
+	 * @param \WC_Product $product The product to check.
+	 * @return bool True if it's a variation, false otherwise.
+	 */
+	private function is_variation( \WC_Product $product ) {
+		return is_a( $product, \WC_Product_Variation::class );
+	}
+
+	/**
+	 * Return the list of taxonomies used for variations on a product together with
+	 * the associated term ids, with the following format:
+	 *
+	 * [
+	 *   'taxonomy_name' =>
+	 *   [
+	 *     'term_ids' => [id, id, ...],
+	 *     'used_for_variations' => true|false
+	 *   ], ...
+	 * ]
+	 *
+	 * @param \WC_Product $product The product to get the attribute taxonomies for.
+	 * @return array Information about the attribute taxonomies of the product.
+	 */
+	private function get_attribute_taxonomies( \WC_Product $product ) {
+		$product_attributes = $product->get_attributes();
+		$result             = array();
+		foreach ( $product_attributes as $taxonomy_name => $attribute_data ) {
+			if ( ! $attribute_data->get_id() ) {
+				// Custom product attribute, not suitable for attribute-based filtering.
+				continue;
+			}
+
+			$result[ $taxonomy_name ] = array(
+				'term_ids'            => $attribute_data->get_options(),
+				'used_for_variations' => $attribute_data->get_variation(),
+			);
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Insert one entry in the lookup table.
+	 *
+	 * @param int    $product_id The product id.
+	 * @param int    $product_or_parent_id The product id for non-variable products, the main/parent product id for variations.
+	 * @param string $taxonomy Taxonomy name.
+	 * @param int    $term_id Term id.
+	 * @param bool   $is_variation_attribute True if the taxonomy corresponds to an attribute used to define variations.
+	 * @param bool   $has_stock True if the product is in stock.
+	 */
+	private function insert_lookup_table_data( int $product_id, int $product_or_parent_id, string $taxonomy, int $term_id, bool $is_variation_attribute, bool $has_stock ) {
+		global $wpdb;
+
+		// phpcs:disable WordPress.DB.PreparedSQL.NotPrepared
```

Comment:
Maybe we don't need the phpcs comment here as this is prepared?

Reply:
Same as in the previous case.